File: crates/dap-script/examples/go/sieve.go
// Sieve of Eratosthenes with deliberately rich state to watch under a debugger:
// a []bool, a nested map[int][]int, a []int, and loop scalars.
package main

import "fmt"

func first_n_primes(n int) []int {
	isPrime := make([]bool, n+1)
	for i := 2; i <= n; i++ {
		isPrime[i] = true
	}

	// crossedOff[p] lists the multiples eliminated by prime p. A map of slices,
	// so the variable tree has nested containers to expand.
	crossedOff := map[int][]int{}

	for p := 2; p*p <= n; p++ {
		if !isPrime[p] {
			continue
		}
		for multiple := p * p; multiple <= n; multiple += p {
			if isPrime[multiple] {
				isPrime[multiple] = false
				crossedOff[p] = append(crossedOff[p], multiple) // good breakpoint: line 26
			}
		}
	}

	primes := []int{}
	for i := 2; i <= n; i++ {
		if isPrime[i] {
			primes = append(primes, i)
		}
	}

	fmt.Println("crossed off:", crossedOff)
	return primes
}

func main() {
	const n = 50
	primes := first_n_primes(n)
	fmt.Println("primes:", primes)
}

File: crates/dap-script/examples/rust/sieve.rs
// Sieve of Eratosthenes with deliberately rich state to watch under a debugger:
// a Vec<bool>, a nested map of vectors, and loop scalars.

use std::collections::HashMap;

fn first_n_primes(n: usize) -> Vec<usize> {
    let mut is_prime = vec![true; n + 1];
    is_prime[0] = false;
    is_prime[1] = false;

    // crossed_off[p] lists the multiples eliminated by prime p. A map of vectors,
    // so the variable tree has nested containers to expand.
    let mut crossed_off: HashMap<usize, Vec<usize>> = HashMap::new();

    let mut p = 2;
    while p * p <= n {
        if is_prime[p] {
            let mut eliminated: Vec<usize> = Vec::new();
            let mut multiple = p * p;
            while multiple <= n {
                if is_prime[multiple] {
                    is_prime[multiple] = false;
                    eliminated.push(multiple); // good breakpoint
                }
                multiple += p;
            }
            if !eliminated.is_empty() {
                crossed_off.insert(p, eliminated);
            }
        }
        p += 1;
    }

    println!("crossed off: {:?}", crossed_off);
    (2..=n).filter(|&i| is_prime[i]).collect()
}

fn main() {
    let primes = first_n_primes(50);
    println!("primes: {:?}", primes);
}
